### ISSUE TITLE
[BiDi] Specify a low-level transport format

### ISSUE THREAD

**Issue by jgraham** [w3c/webdriver#1498]
The bi-directional protocol needs a low-level format for encoding messages that are sent over the wire.

The discussion so far has assumed that the format will be JSON-based; this has compatibility with the HTTP protocol and good human readability. It matches other existing protocols (although note that CDP now has a binary variant) and likely makes sense if we think that the message parsing overhead will be small compared to other latencies in the system.

@burg proposed that we adopt [JSON-RPC](https://www.jsonrpc.org/specification) as the specific flavour of JSON for the protocol. This meets several requirements:
* Provides a command type and a response type with an id field on match commands and responses
* Provides an event type that isn't generated by a command
* Provides an error type

However there are several details that might not match our requirements
* Requires a version number on every message, bloating the protocol for little gain (updating versions in WebDriver would be a difficult undertaking given the compatibility guarantees we provide).
* The protocol has a pipelining feature where an array of commands can be sent at once. This kind of feature seems useful to us, but the particular requirement that the implementation produce an array of responses doesn't seem like it matches use cases we might have (like running multiple concurrent commands and getting the response from each as soon as its ready).
* The spec has several SHOULD-level requirements; this is owrrying from the point of view of obtaining interop.

**Comment by bwalderman** [comment]
@shs96c also had some [feedback](https://lists.w3.org/Archives/Public/public-browser-tools-testing/2020JanMar/0008.html) about the transport format:

- The original wire protocol had numerical status codes. “Nice” to see those are coming back.
  - The “message” in error objects should map to the error strings in the existing spec
  - The “data” in error objects should map to something similar to that in the existing spec.
- We need to understand how to communicate multiple error messages in one response (to maintain strict JSON-RPC compat, only one response message expected per command).
  - Web Inspector protocol may already deviates from this and allows multiple errors (or perhaps they are pasted together into one payload)
- Sending ‘jsonrpc’: “2.0” in every command seems like a waste. Can it be dropped, if we’re already thinking of bending the JSON-RPC protocol…

**Comment by bwalderman** [comment]
> Requires a version number on every message, bloating the protocol for little gain

> Sending ‘jsonrpc’: “2.0” in every command seems like a waste.

Seems like something we can drop, especially if we're only going to support one version of jsonrpc.

> The protocol has a pipelining feature where an array of commands can be sent at once. This kind of feature seems useful to us, but the particular requirement that the implementation produce an array of responses doesn't seem like it matches use cases we might have (like running multiple concurrent commands and getting the response from each as soon as its ready).

JSON-RPC already supports running multiple concurrent commands. If the client sends multiple individual commands without batching, then the server may process and respond to these in any order. There is no requirement that the server acknowledges priors commands before processing the next one. The application itself is left to decide how commands should be ordered and synchronized. It seems like batching is most useful to avoid extra round trips.

> The spec has several SHOULD-level requirements; this is owrrying from the point of view of obtaining interop.

I originally seconded the proposal to use JSON-RPC verbatim and incorporated this resolution into the document, but I've been reconsidering lately and it might be worthwhile to define our own "webdriver flavoured" JSON-RPC protocol. This would allow us to make stronger interop guarantees as well as skip things we don't need like numeric error codes and tagging each message with the protocol version.

**Comment by jgraham** [comment]
>  JSON-RPC already supports running multiple concurrent commands. If the client sends multiple individual commands without batching, then the server may process and respond to these in any order. There is no requirement that the server acknowledges priors commands before processing the next one. The application itself is left to decide how commands should be ordered and synchronized. It seems like batching is most useful to avoid extra round trips.

Yes. One of the differences I see between a typical devtools protocol and a remote automation protocol is that the latter is more often deployed in a way that makes network latency matter. For example running a SauceLabs session puts the network latency between my computer and their data center into the pipeline. In WebDriver/HTTP this usually isn't too bad because the commands are relatively high level, but we've also got some ad-hoc support for batching together various commands (e.g. the actions API which could be understood as ). But if WebDriver/BiDi is going to generally be somewhat lower level we might need more generic mechaisms for reducing the number of round trips. However the mehanism that's in JSON-RPC seems like it's less useful than a simpler mechanism ("send a list of commands and they all get executed and responses come whenever") or a more complex mechanism ("send a list of commands that are executed in order with some way to specify data flow between subsequent commands"), so isn't something we should require.

**Comment by christian-bromann** [comment]
Why can't there be a badge command similar to:

```js
{
    "jsonrpc": "2.0",
    "id": 0,
    "method": "badgeCommand",
    "params": {
        "responseType": "single", // or "complete"
        "commands": [
            {
                "jsonrpc": "2.0",
                "id": 0,
                "method": "getTitle",
                "params": { "browsingContextId": "<ID>" }
            },
            {
                "jsonrpc": "2.0",
                "id": 1,
                "method": "getUrl",
                "params": { "browsingContextId": "<ID>" }
            }
        ]
    }
}
```

with the option to get its responses either one by one using a property like `"responseType": "single"`:

```js
{
    "jsonrpc": "2.0",
    "id": 0,
    "result": [{
        "id": 0,
        "value": { "title": "Example Domain" }
    }]
}
```

or "complete" if only one response is desired:

```js
{
    "jsonrpc": "2.0",
    "id": 0,
    "result": [{
        "id": 0,
        "value": { "title": "Example Domain" }
    }, {
        "id": 1,
        "value": { "title": "https://web.page" }
    }]
}
```

**Comment by jgraham** [comment]
Sure, there can. That, minus the option to get a single response is basically the "simpler mechanism" I referred to above. I think that having a single command that allows sending a list of commands to run without imposing any constraints on the response is the simplest thing we can do to support pipelining, and something that I think should be considered for the initial spec. But if we're doing that we don't also want the low level transport requiring we support a different mechanism with more complex semantics.

**Comment by foolip** [comment]
There was some discussion of this in the [WebDriver/BiDi 2020-04-20 meeting](https://docs.google.com/document/d/1qZcT2RBdmxbzVorV1hTUaH_sIvrhGYjZbP4rK0IpBEc/edit#bookmark=id.608tq39hrexb). Some additional points/considerations not already raised on this thread:
- WebKit's protocol is JSON-RPC-like, with a bunch of tests for the desired behavior, not JSON-RPC as-is
- Size of binary data when encoded as base64 not seen as a big problem by anyone present
- Chromium's CDP uses JSON (over WebSockets) but there's also an [experimental binary protocol](https://bugs.chromium.org/p/chromium/issues/detail?id=929862) that uses [CBOR](https://en.wikipedia.org/wiki/CBOR)

**Comment by jgraham** [comment]
> Chromium's CDP uses JSON (over WebSockets) but there's also an experimental binary protocol that uses CBOR

I think we should start with a JSON transport and consider binary representations if we find that applications are suffering from significant (de)serialization overhead or network traffic. Hopefully it's mostly a drop-in replacement (we'd have some capability in new session to request the alternate encoding, which could either fail if it's not supported or just return the variant that's being used even if that's not the one that was requested. For apps it would presumably just be a case of having an alternate serialization of existing data structures, so not that much work).

**Comment by foolip** [comment]
I've played with some very simple Puppeteer scripts and setting `DEBUG=puppeteer:protocol`(thanks @mathiasbynens) in the environment to learn a bit about what CDP looks like. Here's a sample of what happens when running `await page.evaluate('1 + 1')`:

SEND ►
```json
{
    "id": 17,
    "method": "Runtime.evaluate",
    "params": {
        "awaitPromise": true,
        "contextId": 3,
        "expression": "1 + 1\n//# sourceURL=__puppeteer_evaluation_script__",
        "returnByValue": true,
        "userGesture": true
    },
    "sessionId": "D05EE2DD030AFE4CC6DC5DD4C827F7D7"
}
```

◀ RECV
```
{
    "id": 17,
    "result": {
        "result": {
            "description": "2",
            "type": "number",
            "value": 2
        }
    },
    "sessionId": "D05EE2DD030AFE4CC6DC5DD4C827F7D7"
}
```

In other words, it's a lot like JSON-RPC, but without the `jsonrpc` version stuff, but with a `sessionId` in all messages.

**Comment by foolip** [comment]
> I think we should start with a JSON transport and consider binary representations if we find that applications are suffering from significant (de)serialization overhead or network traffic.

I agree with this. Having skimmed through https://en.wikipedia.org/wiki/CBOR I think there would be some implications for https://github.com/w3c/webdriver/issues/1500 if we went with *only* a binary representation which can represent some things that JSON cannot, but starting with JSON is way to go, I think.

**Comment by foolip** [comment]
Using `safaridriver --diagnose` and doing some things with `safaridriver` one can also have a look at what the underlying protocol looks like from `~/Library/Logs/com.apple.WebDriver/`. This is from running one test in WPT:

AutomationProtocol: SEND
```json
{
    "id": 28,
    "method": "Automation.waitForNavigationToComplete",
    "params": {
        "browsingContextHandle": "page-FD3132E2-2ECC-4C2A-B203-5125CEB5AD80",
        "frameHandle": "",
        "pageLoadStrategy": "Eager",
        "pageLoadTimeout": 300000
    }
}
```

AutomationProtocol: RECV
```
{
    "id": 28,
    "result": {}
}
```

Very JSON-RPC-like indeed. (Sorry for the big comments, but thought it'd be good to document in a hands-on way what things look like so there's no doubt.)

**Comment by foolip** [comment]
I was curious about `sessionId` in CDP and did some protocol testing by starting Chrome with `--remote-debugging-port=9999` and then connecting to one of the WebSocket URLs listed by http://localhost:9999/json/version.

`sessionId` isn't required, the smallest command I could come up with for   `{"id":0,"method":"Runtime.evaluate","params":{"expression":"1 + 1"}}`, which gets the response `{"id":0,"result":{"result":{"type":"number","value":2,"description":"2"}}}`.

**Comment by foolip** [comment]
So, after that digging, I think the starting point should be https://www.jsonrpc.org/specification without the `jsonrpc` members. To summarize:
- client&rarr;browser commands are objects with `method` (string), `params` (object) and `id` (integer, more strict that JSON-RPC)
- browser&rarr;client responses are objects with `result` (always object?) and `id` (integer)
- browser&rarr;client events are objects with `method` (string), `params` (optional?) and *no* `id`
- browser&rarr;client errors are like response objects but with an `error` member.

There are details to work out, but I think that's enough to try fleshing it out and discover what's wrong.

**Comment by foolip** [comment]
All of the above discussion concerns the message format, but one layer below that we also need the protocol for exchanging messages in the first place. The [WebSocket protocol](https://tools.ietf.org/html/rfc6455) is a natural contender for this, and is what Chromium uses.

@jgraham @burg how about Gecko and WebKit, is WebSockets used in a similar way there, or is it something else?

**Comment by jgraham** [comment]
I think there was previously a decision to use WebSockets here, or at least we discussed in those terms and it wasn't controversial.

**Comment by bwalderman** [comment]
A `sessionId` is included in CDP messages when using the protocol in "flat" mode. This allows you to use a single underlying WebSocket connection to talk to multiple Targets. The `sessionId` indicates which Target a message should be routed to.

The current proposal is to use a single WebSocket per bidi WebDriver session, so we'd need something similar to route commands to the correct target. This information (e.g. `browsingContextId` or `frameId`) could be included in a command's `params` without needing to change the transport format.

> client→browser commands are objects with method (string), params (object) and id (integer, more strict that JSON-RPC)
> browser→client responses are objects with result (always object?) and id (integer)
> browser→client events are objects with method (string), params (optional?) and no id
> browser→client errors are like response objects but with an error member.

Sounds good to me. I would add that `error` objects should use the [WebDriver](https://w3c.github.io/webdriver/#errors) format instead of the JSON-RPC format (descriptive string error codes instead of integer error codes).
